File: app/src/main/java/com/tunjid/androidbootstrap/fragments/DoggoPagerFragment.java
package com.tunjid.androidbootstrap.fragments;

import android.annotation.SuppressLint;
import android.content.res.Resources;
import android.os.Bundle;
import android.transition.Transition;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.tunjid.androidbootstrap.R;
import com.tunjid.androidbootstrap.adapters.DoggoPagerAdapter;
import com.tunjid.androidbootstrap.baseclasses.AppBaseFragment;
import com.tunjid.androidbootstrap.core.abstractclasses.BaseFragment;
import com.tunjid.androidbootstrap.model.Doggo;
import com.tunjid.androidbootstrap.view.animator.FabExtensionAnimator.GlyphState;
import com.tunjid.androidbootstrap.view.animator.ViewPagerIndicatorAnimator;
import com.tunjid.androidbootstrap.view.util.InsetFlags;
import com.tunjid.androidbootstrap.view.util.ViewUtil;

import java.util.List;
import java.util.Map;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.app.SharedElementCallback;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.viewpager.widget.ViewPager;
import androidx.viewpager.widget.ViewPager.SimpleOnPageChangeListener;

import static com.tunjid.androidbootstrap.view.animator.FabExtensionAnimator.newState;

public class DoggoPagerFragment extends AppBaseFragment {

    private ViewPager viewPager;

    public static DoggoPagerFragment newInstance() {
        DoggoPagerFragment fragment = new DoggoPagerFragment();
        fragment.setArguments(new Bundle());
        fragment.prepareSharedElementTransition();
        return fragment;
    }

    @Nullable
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_doggo_pager, container, false);
        Resources resources = getResources();
        int indicatorSize = resources.getDimensionPixelSize(R.dimen.single_and_half_margin);

        viewPager = root.findViewById(R.id.view_pager);
        viewPager.setAdapter(new DoggoPagerAdapter(Doggo.doggos, getChildFragmentManager()));
        viewPager.setCurrentItem(Doggo.getTransitionIndex());
        viewPager.addOnPageChangeListener(new SimpleOnPageChangeListener() {
            public void onPageSelected(int position) { onDoggoSwiped(position); }
        });

        ViewPagerIndicatorAnimator indicatorAnimator = ViewPagerIndicatorAnimator.builder()
                .setIndicatorWidth(indicatorSize)
                .setIndicatorHeight(indicatorSize)
                .setIndicatorPadding(resources.getDimensionPixelSize(R.dimen.half_margin))
                .setInActiveDrawable(R.drawable.ic_circle_24dp)
                .setActiveDrawable(R.drawable.ic_doggo_24dp)
                .setGuideLine(root.findViewById(R.id.guide))
                .setContainer((ConstraintLayout) root)
                .setViewPager(viewPager)
                .build();

        indicatorAnimator.addIndicatorWatcher((indicator, position, fraction, totalTranslation) -> {
            double radians = Math.PI * fraction;
            float sine = (float) -Math.sin(radians);
            float cosine = (float) Math.cos(radians);
            float maxScale = Math.max(Math.abs(cosine), 0.4F);

            ImageView currentIndicator = indicatorAnimator.getIndicatorAt(position);
            currentIndicator.setScaleX(maxScale);
            currentIndicator.setScaleY(maxScale);
            indicator.setTranslationY(indicatorSize * sine);
        });

        prepareSharedElementTransition();
        tintView(R.color.black, root, (color, view) -> view.setBackgroundColor(color));
        if (savedInstanceState == null) postponeEnterTransition();
        return root;
    }

    private void onDoggoSwiped(int position) {
        Doggo.setTransitionDoggo(Doggo.doggos.get(position));
        togglePersistentUi();
    }

    public InsetFlags insetFlags() { return InsetFlags.NONE; }

    public boolean showsToolBar() { return false; }

    public boolean showsFab() { return true; }

    public GlyphState getFabState() {
        return newState(getDogName(), ContextCompat.getDrawable(requireContext(), R.drawable.ic_hug_24dp));
    }

    @Override
    protected View.OnClickListener getFabClickListener() {
        return view -> showFragment(AdoptDoggoFragment.newInstance(Doggo.getTransitionDoggo()));
    }

    @Nullable
    @SuppressLint({"CommitTransaction"})
    public FragmentTransaction provideFragmentTransaction(BaseFragment to) {
        if (!to.getStableTag().contains(AdoptDoggoFragment.class.getSimpleName())) return null;

        View root = getView();
        if (root == null) return null;

        Doggo doggo = Doggo.getTransitionDoggo();
        if (doggo == null) return null;

        View childRoot = root.findViewWithTag(doggo);
        if (childRoot == null) return null;

        ImageView imageView = childRoot.findViewById(R.id.doggo_image);
        if (imageView == null) return null;

        return requireActivity()
                .getSupportFragmentManager()
                .beginTransaction()
                .setReorderingAllowed(true)
                .addSharedElement(imageView, ViewUtil.transitionName(doggo, imageView));
    }

    private String getDogName() {
        Doggo doggo = Doggo.getTransitionDoggo();
        if (doggo == null) return "";

        String name = doggo.getName().replace(" ", "");
        return getString(R.string.adopt_doggo, name);
    }

    private void prepareSharedElementTransition() {
        Transition baseTransition = baseTransition();
        Transition baseSharedTransition = baseSharedTransition();

        setEnterTransition(baseTransition);
        setExitTransition(baseTransition);
        setSharedElementEnterTransition(baseSharedTransition);
        setSharedElementReturnTransition(baseSharedTransition);

        setEnterSharedElementCallback(new SharedElementCallback() {
            public void onMapSharedElements(List<String> names, Map<String, View> sharedElements) {
                Fragment currentFragment = (Fragment) Objects.requireNonNull(viewPager.getAdapter()).instantiateItem(viewPager, Doggo.getTransitionIndex());
                View view = currentFragment.getView();
                if (view == null) return;

                sharedElements.put(names.get(0), view.findViewById(R.id.doggo_image));
            }
        });
    }
}

File: app/src/main/java/com/tunjid/androidbootstrap/fragments/DoggoPagerFragment.kt
package com.tunjid.androidbootstrap.fragments

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.widget.ImageView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.app.SharedElementCallback
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import androidx.fragment.app.viewModels
import androidx.lifecycle.observe
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.SimpleOnPageChangeListener
import com.tunjid.androidbootstrap.GlobalUiController
import com.tunjid.androidbootstrap.R
import com.tunjid.androidbootstrap.UiState
import com.tunjid.androidbootstrap.activityGlobalUiController
import com.tunjid.androidbootstrap.adapters.DoggoPagerAdapter
import com.tunjid.androidbootstrap.baseclasses.AppBaseFragment
import com.tunjid.androidbootstrap.constraintlayout.animator.ViewPagerIndicatorAnimator
import com.tunjid.androidbootstrap.core.components.FragmentStackNavigator
import com.tunjid.androidbootstrap.model.Doggo
import com.tunjid.androidbootstrap.view.util.InsetFlags
import com.tunjid.androidbootstrap.view.util.ViewUtil
import com.tunjid.androidbootstrap.viewmodels.DoggoViewModel
import java.util.*
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.sin

class DoggoPagerFragment : AppBaseFragment(R.layout.fragment_doggo_pager), GlobalUiController {

    override var uiState: UiState by activityGlobalUiController()

    private val viewModel by viewModels<DoggoViewModel>()

    override val insetFlags: InsetFlags = InsetFlags.NONE

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel.getColors(Color.TRANSPARENT).observe(this) { view?.setBackgroundColor(it) }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        uiState = uiState.copy(
                toolbarShows = false,
                toolBarMenu = 0,
                fabIcon = R.drawable.ic_hug_24dp,
                fabShows = true,
                fabExtended = !restoredFromBackStack(),
                showsBottomNav = false,
                navBarColor = Color.TRANSPARENT,
                fabClickListener = View.OnClickListener { Doggo.transitionDoggo?.let { navigator.show(AdoptDoggoFragment.newInstance(it)) } }
        )

        val viewPager = view.findViewById<ViewPager>(R.id.view_pager)
        val resources = resources
        val indicatorSize = resources.getDimensionPixelSize(R.dimen.single_and_half_margin)

        viewPager.adapter = DoggoPagerAdapter(viewModel.doggos, childFragmentManager)
        viewPager.currentItem = Doggo.transitionIndex
        viewPager.addOnPageChangeListener(object : SimpleOnPageChangeListener() {
            override fun onPageSelected(position: Int) = onDoggoSwiped(position)
        })

        val indicatorAnimator = ViewPagerIndicatorAnimator.builder()
                .setIndicatorWidth(indicatorSize)
                .setIndicatorHeight(indicatorSize)
                .setIndicatorPadding(resources.getDimensionPixelSize(R.dimen.half_margin))
                .setInActiveDrawable(R.drawable.ic_circle_24dp)
                .setActiveDrawable(R.drawable.ic_doggo_24dp)
                .setGuideLine(view.findViewById(R.id.guide))
                .setContainer(view as ConstraintLayout)
                .setViewPager(viewPager)
                .build()

        indicatorAnimator.addIndicatorWatcher { indicator, position, fraction, _ ->
            val radians = Math.PI * fraction
            val sine = (-sin(radians)).toFloat()
            val cosine = cos(radians).toFloat()
            val maxScale = max(abs(cosine), 0.4f)

            val currentIndicator = indicatorAnimator.getIndicatorAt(position)
            currentIndicator.scaleX = maxScale
            currentIndicator.scaleY = maxScale
            indicator.translationY = indicatorSize * sine
        }

        indicatorAnimator.addIndicatorWatcher watcher@{ indicator, position, fraction, _ ->
            if (fraction == 0F) return@watcher

            val static = intArrayOf(0, 0).apply { indicatorAnimator.getIndicatorAt(position).getLocationInWindow(this) }
            val dynamic = intArrayOf(0, 0).apply { indicator.getLocationInWindow(this) }
            val toTheRight = dynamic[0] > static[0]

            viewModel.onSwiped(position, fraction, toTheRight)
        }

        onDoggoSwiped(viewPager.currentItem)
        prepareSharedElementTransition()

        if (savedInstanceState == null) postponeEnterTransition()
    }

    private fun onDoggoSwiped(position: Int) {
        Doggo.doggos[position].apply {
            Doggo.transitionDoggo = this
            uiState = uiState.copy(fabText = getString(R.string.adopt_doggo, name.replace(" ", "")))
        }
    }

    @SuppressLint("CommitTransaction")
    override fun augmentTransaction(transaction: FragmentTransaction, incomingFragment: Fragment) {
        if (incomingFragment !is FragmentStackNavigator.TagProvider) return
        if (!incomingFragment.stableTag.contains(AdoptDoggoFragment::class.java.simpleName)) return

        val root = view ?: return

        val doggo = Doggo.transitionDoggo ?: return

        val childRoot = root.findViewWithTag<View>(doggo) ?: return

        val imageView = childRoot.findViewById<ImageView>(R.id.doggo_image) ?: return

        transaction
                .setReorderingAllowed(true)
                .addSharedElement(imageView, ViewUtil.transitionName(doggo, imageView))
    }

    private fun prepareSharedElementTransition() {
        val baseSharedTransition = baseSharedTransition()

        sharedElementEnterTransition = baseSharedTransition
        sharedElementReturnTransition = baseSharedTransition

        setEnterSharedElementCallback(object : SharedElementCallback() {
            override fun onMapSharedElements(names: List<String>?, sharedElements: MutableMap<String, View>?) {
                val viewPager = view?.findViewById<ViewPager>(R.id.view_pager) ?: return
                if (names == null || sharedElements == null || view == null) return

                val currentFragment = Objects.requireNonNull<PagerAdapter>(viewPager.adapter)
                        .instantiateItem(viewPager, Doggo.transitionIndex) as Fragment
                val view = currentFragment.view ?: return

                sharedElements[names[0]] = view.findViewById(R.id.doggo_image)
            }
        })
    }

    companion object {
        fun newInstance(): DoggoPagerFragment = DoggoPagerFragment().apply { arguments = Bundle(); prepareSharedElementTransition() }
    }

}
